server/src/endpoints/store.rs

```diff
@@ -44,19 +44,22 @@ pub enum BadStoreRequest {
     #[fail(display = "failed to read request body")]
     PayloadError(#[cause] StorePayloadError),
 
-    #[fail(display = "event rejected due to rate limit ({}s)", _0)]
-    RateLimited(u64),
+    #[fail(
+        display = "event rejected due to rate limit ({}s) with reason:{}",
+        _0, _1
+    )]
+    RateLimited(u64, String),
 
-    #[fail(display = "event submission rejected")]
-    EventRejected,
+    #[fail(display = "event submission rejected with_reason:{:?}", _0)]
+    EventRejected(DiscardReason),
 }
 
 impl ResponseError for BadStoreRequest {
     fn error_response(&self) -> HttpResponse {
         let body = ApiErrorResponse::from_fail(self);
 
         match self {
-            BadStoreRequest::RateLimited(secs) => {
+            BadStoreRequest::RateLimited(secs, _reason) => {
```

Comment:
Can we add this reason to a response header, like `X-Sentry-Error` or `Warning`?